state/upgrades.go

```diff
@@ -711,3 +712,100 @@ func AddStatusHistoryPruneSettings(st *State) error {
 	}
 	return nil
 }
+
+// AddStorageInstanceConstraints sets the "constraints" field on
+// storage instance docs.
+func AddStorageInstanceConstraints(st *State) error {
+	return runForAllModelStates(st, addStorageInstanceConstraints)
+}
+
+func addStorageInstanceConstraints(st *State) error {
+	storageInstances, err := st.storageInstances(bson.D{
+		{"constraints", bson.D{{"$exists", false}}},
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var ops []txn.Op
+	for _, s := range storageInstances {
+		var siCons storageInstanceConstraints
+		var defaultPool string
+		switch s.Kind() {
+		case StorageKindBlock:
+			v, err := st.storageInstanceVolume(s.StorageTag())
+			if err == nil {
+				if v.doc.Info != nil {
+					siCons.Pool = v.doc.Info.Pool
+					siCons.Size = v.doc.Info.Size
+				} else if v.doc.Params != nil {
+					siCons.Pool = v.doc.Params.Pool
+					siCons.Size = v.doc.Params.Size
+				}
+			} else if errors.IsNotFound(err) {
+				defaultPool = string(provider.LoopProviderType)
+			} else {
+				return errors.Trace(err)
+			}
+		case StorageKindFilesystem:
+			f, err := st.storageInstanceFilesystem(s.StorageTag())
+			if err == nil {
+				if f.doc.Info != nil {
+					siCons.Pool = f.doc.Info.Pool
+					siCons.Size = f.doc.Info.Size
+				} else if f.doc.Params != nil {
+					siCons.Pool = f.doc.Params.Pool
+					siCons.Size = f.doc.Params.Size
+				}
+			} else if errors.IsNotFound(err) {
+				defaultPool = string(provider.RootfsProviderType)
+			} else {
+				return errors.Trace(err)
+			}
+		default:
+			// Unknown storage kind, ignore.
+			continue
+		}
+		if siCons.Pool == "" {
+			// There's no associated volume or filesystem, so
```

Comment:
That is true; it's more wishful thinking that we had some way of surfacing to the user as output of the upgrade-juju command any such issues they should be aware of. At the least we could log so that there's at least some record of it. They can't necessarily do anything easy but it explicitly informs them that some assumptions they may have had are not valid and removes the surprise element.